File: examples/payment-service-go/retry.go
package payment

import (
	"fmt"
	"time"
)

type RetryConfig struct {
	MaxAttempts int
	Delay       time.Duration
}

func WithRetry(fn func() (string, error), cfg RetryConfig) (string, error) {
	var lastErr error

	for attempt := 0; attempt < cfg.MaxAttempts; attempt++ {
		result, err := fn()
		if err == nil {
			return result, nil
		}

		lastErr = err

		if attempt < cfg.MaxAttempts-1 {
			time.Sleep(cfg.Delay)
		}
	}

	return "", fmt.Errorf("payment failed after %d attempts: %w", cfg.MaxAttempts, lastErr)
}

File: examples/payment-service-rust/src/retry.rs
use std::time::Duration;
use tokio::time::sleep;

pub struct RetryConfig {
    pub max_attempts: u32,
    pub delay: Duration,
}

pub async fn with_retry<F, Fut, T, E>(config: &RetryConfig, mut f: F) -> Result<T, E>
where
    F: FnMut() -> Fut,
    Fut: std::future::Future<Output = Result<T, E>>,
    E: std::fmt::Debug,
{
    let mut attempts = 0;
    let mut last_err = None;

    loop {
        match f().await {
            Ok(value) => return Ok(value),
            Err(e) => {
                last_err = Some(e);
                attempts += 1;

                if attempts >= config.max_attempts {
                    break;
                }

                if attempts < config.max_attempts {
                    sleep(config.delay).await;
                }
            }
        }
    }

    Err(last_err.expect("loop exited without error"))
}
